Tag private packages when their version changes

JakeGinnivan opened the pull request:
Replaces #420 and #569. Also addresses comment in https://github.com/atlassian/changesets/issues/478#issuecomment-780951976

This updates the publishPackage function so it returns 2 things, the publishedPackages and the untaggedPrivatePackages. I wasn't sure how far to take the separation, it could have been it's own module.

Possibly a further step would be to gather up the `projects` which need to be published & tagged. This would allow changesets to be extended to publish .NET packages to NuGet for instance.

This could also be considered as a breaking change. We possibly should add a new config option to opt into this?

JakeGinnivan commented:
@gakonst @Andarist @mitchellhamilton here is another attempt, I think it addresses (or is closer to addressing) Mitchell's feedback on the other approaches.

JakeGinnivan commented:
With this approach we would need to fetch tags as well, otherwise it will continually trigger releases

vemundeldegard commented:
I would love this feature for tagging private packages. I thought this would come out of the box with Changesets, but sad to see that it still isn't possible...

Andarist commented:
@vemundeldegard could you describe your use case? With more data we’ll be able to take a better/more informed final decision about this

vemundeldegard commented:
> @vemundeldegard could you describe your use case? With more data we’ll be able to take a better/more informed final decision about this

Sure! 

In our project we have a couple of private packages and a couple of public packages. The private packages are basically a handful of AWS Lambdas. We want to use Changesets because it offers a handy developer experience for managing versioning and changelog. My understanding is that Changesets do update and merge the changelogs for private our private packages, but it does not create a new tag og a GitHub release. Please correct me if this is not right.. 

The above is the main drawback for us, because we use the releases to trigger additional steps in our workflow, such as deploying the Lambdas. It has forced us to only use the `changeset` command, and `changeset version`. Right now we are experimenting with using `release-it` to handle the tagging and releasing when a changeset PR is merged.

JakeGinnivan commented:
@Andarist should I rebase this? It incorporates the feedback from @mitchellhamilton has for the previous 3 attempts

JakeGinnivan commented:
I am using this via patch-package locally. As for why, I use it to control the releases of featureboard.app to production and manage my changelogs.

<img width="811" alt="image" src="https://user-images.githubusercontent.com/453152/152251600-e735e386-e8b1-46e3-8b1d-d376bfb70ae5.png">

miszo commented:
This looks neat and would address the question asked by me in #745 🙂

JakeGinnivan commented:
> This looks neat and would address the question asked by me in #745 🙂

@miszo Any feedback on the implementation? I think @Andarist is after someone to critically review who needs this feature. I will rebase do at least there are not conflicts to start

miszo commented on packages/cli/src/commands/publish/index.ts:
```diff
@@ -66,36 +66,58 @@ export default async function run(
 
-  if (unsuccessful.length > 0) {
+  if (untaggedPrivatePackages.length > 0) {
+    success("found untagged projects:");
+    logReleases(untaggedPrivatePackages);
+
+    if (tool !== "root") {
+      for (const pkg of untaggedPrivatePackages) {
+        const tag = `${pkg.name}@${pkg.newVersion}`;
+        log("New tag: ", tag);
+        await git.tag(tag, cwd);
+      }
+    } else {
+      const tag = `v${untaggedPrivatePackages[0].newVersion}`;
+      log("New tag: ", tag);
+      await git.tag(tag, cwd);
+    }
+  }
```
I guess that this logic is duplicated and could be extracted.

JakeGinnivan replied:
Yeah it is similar, I might leave it for simplicity

miszo reviewed (approved):
@JakeGinnivan, I guess that besides the little duplication, the implementation looks good.

pradel commented:
I would also be interested to see this one released.
For the use case, I am managing an open-source application where we have apps pushed to docker hub and packages released to npm.
I want to integrate changeset in the repository to manage the versionning + changeset of the apps published to docker as well as the packages. The docker images are built and pushed using a github action listening to git tags pushed to the main branch. The issue we have right now is that the apps are not creating a tag so the GitHub action is not triggered.

JakeGinnivan commented:
Rebased on main.

miszo commented:
@JakeGinnivan, @Andarist – can we proceed and merge this pull request? Are there any blockers or other reasons that suggests that this change shouldn't be merged?

JakeGinnivan commented:
I've rebased again, I am referencing the packages built by codesandbox directly in a number of projects. 
But when I rebase they 404 and I have to update all my projects, would love to get this merged.

Andarist commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -112,6 +113,52 @@ export default async function publishPackages({
+  const packageWithTags = await Promise.all(
+    privatePackages.map(async privatePkg => {
+      const tagName = `${privatePkg.packageJson.name}@${privatePkg.packageJson.version}`;
+      const isMissingTag = !(await git.tagExists(tagName));
```
q: shouldn't we query a remote here?

JakeGinnivan replied:
I am adding documentation and saying the user should fetch tags before running the `publish` command. I don't want to introduce the tool doing fetches, I think that should stay outside the CLI

JakeGinnivan replied:
Fixed, also means you no longer have to pull down all the tags locally which saves on bandwidth

Andarist commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -112,6 +113,52 @@ export default async function publishPackages({
+async function getUntaggedPrivatePackages(privatePackages: Package[]) {
+  const packageWithTags = await Promise.all(
+    privatePackages.map(async privatePkg => {
+      const tagName = `${privatePkg.packageJson.name}@${privatePkg.packageJson.version}`;
```
I think that it is not guaranteed that a private package will have a version field - we should account for that

JakeGinnivan replied:
I am now filtering packages which do not have a `version` field, added this to the doco I am writing.

Andarist commented:
I think that:
- we don't have to distinguish between private and public packages externally, only `publishPackages` has to care about this, its return value shouldn't though
- we might want to create those tags conditionally (to not change this behavior in a minor version). We should either introduce a config flag or a CLI flag for this. Do you have any preference as to which it should be?

JakeGinnivan commented:
> we don't have to distinguish between private and public packages externally, only publishPackages has to care about this, its return value shouldn't though

Eventually I think this function should not actually publish, instead return the packages which *should* be published then have multiple publish strategies like publish to NPM, run command etc.

This will allow these private packages to maybe have a `release` script which can be invoked by changesets to do a custom release (to gems or something).

> we might want to create those tags conditionally (to not change this behavior in a minor version). We should either introduce a config flag or a CLI flag for this. Do you have any preference as to which it should be?

Added a config flag `enablePrivatePackageTracking`

JakeGinnivan commented:
> we don't have to distinguish between private and public packages externally, only publishPackages has to care about this, its return value shouldn't though

Further to this, tagging is done outside of the publishPackages function. So I need to return the unpublished private packages so I don't need to duplicate the tagging code

JakeGinnivan commented:
@Andarist have addressed feedback, rebased, added some doco and also implemented the changes in https://github.com/changesets/changesets/pull/720 which takes the config value into account so it will also fix #620

Andarist commented on packages/git/src/index.test.ts:
```diff
@@ -226,6 +227,23 @@ describe("git", () => {
   });
 
+  describe("tagExists", () => {
+    beforeEach(async () => {
+      await add("packages/pkg-a/package.json", cwd);
+      await commit("added packageA package.json", cwd);
+    });
+
+    it("returns false when no tag exists", async () => {
+      expect(await tagExists("tag_name")).toBe(false);
+    });
+
+    it("should create a tag for the current head", async () => {
+      await tag("tag_message", cwd);
+
+      expect(await tagExists("tag_name")).toBe(false);
+    });
```
```suggestion
    it("returns false when no tag exists", async () => {
      await add("packages/pkg-a/package.json", cwd);
      await commit("added packageA package.json", cwd);
      expect(await tagExists("tag_name")).toBe(false);
    });

    it("should create a tag for the current head", async () => {
      await add("packages/pkg-a/package.json", cwd);
      await commit("added packageA package.json", cwd);
      await tag("tag_message", cwd);

      expect(await tagExists("tag_name")).toBe(false);
    });
```

JakeGinnivan commented on docs/versioning-apps.md:
```diff
@@ -0,0 +1,25 @@
+
+The only requirement is that the project has a package.json file to manage the versions and dependencies within the repo.
+
+To enable this feature set `enablePrivatePackageTracking` to `true` in your `.changesets/config.json` file.
```
Yeah, maybe 'enableApplicationTracking' or similar. It would be a keyword match for the doc at least?

JakeGinnivan commented on docs/versioning-apps.md:
```diff
@@ -0,0 +1,25 @@
+> Changesets only versions NPM package.json files, you can trigger releases for other package formats by creating workflows which trigger on tags/releases being created by changesets.
+
+> **Note**
+> This feature relies on git tags to track releases, ensure you fetch tags before you publish.
```
Not particularly, it is just a change which can increase traffic/slow down builds. Maybe create an opt out in config?

JakeGinnivan replied:
Have enabled checking remote tags

JakeGinnivan commented on packages/cli/src/commands/add/createChangeset.ts:
```diff
@@ -58,6 +59,20 @@ async function getPackagesToRelease(
 
+  const pkgJsonsByName = getPkgJsonByName(allPackages);
+
+  // filter out packages which changesets is not tracking
+  allPackages = allPackages.filter(pkg =>
+    isValidChangesetPackage(pkg.packageJson, trackPrivatePackages)
+  );
+  changedPackages = changedPackages.filter(
+    pkgName =>
+      !isValidChangesetPackage(
+        pkgJsonsByName.get(pkgName)!,
+        trackPrivatePackages
+      )
+  );
```
Interesting. Maybe have a `privatePackages: 'track' | 'version'` or similar rather than the config flag? I'm worried about complexity though but would allow us to introduce new modes easily?

JakeGinnivan replied:
This now only filters if privatePackages === 'ignore'

JakeGinnivan commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -78,30 +79,41 @@ export default async function publishPackages({
   tag?: string;
-}) {
+  trackPrivatePackages: boolean;
+}): Promise<{
+  publishedPackages: PublishedResult[];
+  untaggedPrivatePackages: PublishedResult[];
+}> {
```
I guess because I feel the logic is a little broken, but it would require more extensive refactoring to fix.

publishPackages both figures out what packages need to publish and also publishes them. It seemed like the best spot to keep the similar logic of figuring out unpublished private packages too.

I kinda feel that eventually this logic will be pulled apart so Changesets has 2 distinct phases, first figuring out what needs to be published, then the publish step.

This would allow publishing to other package managers and make changesets much less tied to NPM.

Also felt a little cleaner separating them, I probably should have used different types because it isn't a PublishedResult.

xiaoyu-tamu commented:
any idea when will this PR get merged?

chathu-novade commented:
Any update on this?

Andarist commented on packages/config/src/index.ts:
```diff
@@ -460,8 +461,18 @@ export let parse = (json: WrittenConfig, packages: Packages): Config => {
+    },
+
+    // TODO consider enabling this by default in the next major version
+    privatePackages:
+      json.privatePackages === "ignore"
+        ? PrivatePackages.Ignore
+        : json.privatePackages === "version-and-tag"
+        ? PrivatePackages.VersionAndTag
+        : // Default value
+          PrivatePackages.Version
```
I'd probably keep the public and internal representation as close as possible. The internal one in most cases should be the normalized public value where that normalized value is also a valid public value. I feel like having those look differently creates an unnecessary barrier/indirection

Andarist commented on packages/types/src/index.ts:
```diff
@@ -61,13 +61,26 @@ export type PackageGroup = ReadonlyArray<string>;
+ */
+export const PrivatePackages = {
+  Ignore: 0,
+  Version: 1,
```
is this utilized anywhere? I can't find it anywhere in the code

Andarist replied:
Oh, I think that it might be used indirectly through `.Ignore`. nit: with bit flags check I didn't expect that, the code depending on the version flag should only check if that bit is present, not that there are no bits enabled at all

Andarist commented on packages/cli/src/commands/add/createChangeset.ts:
```diff
@@ -92,29 +107,44 @@ async function getPackagesToRelease(
   );
 }
 
+function isValidChangesetPackage(
```
nit:
```suggestion
function canPackageBeVersioned(
```

Andarist commented on packages/cli/src/commands/add/createChangeset.ts:
```diff
@@ -58,6 +59,20 @@ async function getPackagesToRelease(
 
+  const pkgJsonsByName = getPkgJsonsByName(allPackages);
+
+  // filter out packages which changesets is not tracking
```
it seems that we have a similar filter in the function that calls this one (see [here](https://github.com/changesets/changesets/blob/98d63e0a85506175e219b93478a814f8588c52d3/packages/cli/src/commands/add/index.ts#L18-L23)). Should those places be merged?

Andarist commented on packages/cli/src/commands/add/index.ts:
```diff
@@ -48,7 +48,11 @@ export default async function add(
+    newChangeset = await createChangeset(
+      changedPackagesName,
+      packages,
+      config.privatePackages === PrivatePackages.Ignore
```
As mentioned elsewhere, with bit flags I'd expect smth like here
```suggestion
      !(config.privatePackages & PrivatePackages.Version)
```

Andarist commented on packages/cli/src/commands/publish/index.ts:
```diff
@@ -63,39 +63,69 @@ export default async function run(
 }
+
+function isFlagEnabled(value: number, tag: number): boolean {
+  return (value & tag) === tag;
```
nit: that could sometimes be a valid check to do, to check if the tag is **exactly** the same as the bitmask - but usually we just want to check `& tag`

JakeGinnivan replied:
Not using flags now

Andarist commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -78,30 +79,41 @@ export default async function publishPackages({
+  tagPrivatePackages: boolean;
+}): Promise<{
+  publishedPackages: PublishedResult[];
+  untaggedPrivatePackages: Omit<PublishedResult, "published">[];
```
I think that for now, I would prefer to hoist handling of the taggable private packages one level up. Let this `publishPackages` function only be responsible for npm-related packages.

We should group packages that one level up and only pass public packages to this function here - while simultaneously checking for the remove tags of the private packages.

JakeGinnivan replied:
Will have a fresh crack at cleaning this up. I avoided restructuring where things happened too much in this PR but it has made it harder to follow

JakeGinnivan replied:
Ok, I think I did it this way because of the publish warn if there is no packages to publish. This warn also has been hoisted.

This looks nicer.

Andarist commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -112,6 +124,52 @@ export default async function publishPackages({
+  const packageWithTags = await Promise.all(
+    privatePackages.map(async privatePkg => {
+      const tagName = `${privatePkg.packageJson.name}@${privatePkg.packageJson.version}`;
+      const isMissingTag = !(await git.remoteTagExists(tagName));
```
I wonder if it would be worth doing this:
```suggestion
      const isMissingTag = !(await git.tagExists(tagName) || await git.remoteTagExists(tagName));
```
WDYT?

JakeGinnivan replied:
Good idea

JakeGinnivan commented on packages/types/src/index.ts:
```diff
@@ -90,6 +103,8 @@ export type WrittenConfig = {
   access?: AccessType;
   baseBranch?: string;
+  /** Opt in to tracking non-npm / private packages */
+  privatePackages?: "ignore" | "version-without-tag" | "version-and-tag";
```
I don't have strong feels, I don't mind your suggestion and it would be pretty easy to use a union type to represent the valid combinations?

Will try, it also will cleanup the flags stuff, which I think having an object like this may be simpler to maintain and people to understand in the long run.

JakeGinnivan replied:
Done

JakeGinnivan commented on packages/cli/src/commands/add/isListablePackage.ts:
```diff
@@ -0,0 +1,17 @@
+import { Config } from "@changesets/types";
+import { PackageJSON } from "@changesets/types";
+
+export function isListablePackage(config: Config, packageJson: PackageJSON) {
```
@Andarist I went with longhand logic here rather than the terse version because it's now more complicated

JakeGinnivan commented on packages/git/src/index.test.ts:
```diff
@@ -226,6 +227,22 @@ describe("git", () => {
+      expect(await tagExists("tag_name")).toBe(false);
+    });
+
+    it("should create a tag for the current head", async () => {
+      await add("packages/pkg-a/package.json", cwd);
+      await commit("added packageA package.json", cwd);
+      await tag("tag_message", cwd);
+
+      expect(await tagExists("tag_name")).toBe(false);
+    });
```
Great catch, added cwd to bring into line with other commands

Andarist commented on .changeset/fast-jars-thank.md:
```diff
@@ -0,0 +1,9 @@
+---
```
this should be adjusted according to the latest changes

Andarist commented on docs/versioning-apps.md:
```diff
@@ -0,0 +1,22 @@
+
+The only requirement is that the project has a package.json file to manage the versions and dependencies within the repo.
+
+To enable this feature set `privatePackages` to `version-and-tag` in your `.changesets/config.json` file. By default changesets will only update the changelog and version
```
This also needs to be adjusted according to the latest changes.

Btw. thank you very much for including docs in your PR ❤️ ❤️ ❤️

Andarist commented on packages/apply-release-plan/src/index.test.ts:
```diff
@@ -49,6 +49,7 @@ class FakeReleasePlan {
       baseBranch: "main",
       updateInternalDependencies: "patch",
       ignore: [],
+      privatePackages: { version: true, tag: false },
```
It reminds me that we need to also update the JSON schema for the config object:
https://github.com/changesets/changesets/blob/98d63e0a85506175e219b93478a814f8588c52d3/packages/config/schema.json

JakeGinnivan replied:
Nice catch

Andarist commented on packages/cli/src/commands/add/createChangeset.ts:
```diff
@@ -58,6 +60,16 @@ async function getPackagesToRelease(
+  const pkgJsonsByName = getPkgJsonsByName(allPackages);
+
+  // filter out packages which changesets is not tracking
+  allPackages = allPackages.filter(pkg =>
+    isListablePackage(config, pkg.packageJson)
+  );
+  changedPackages = changedPackages.filter(
+    pkgName => !isListablePackage(config, pkgJsonsByName.get(pkgName)!)
+  );
```
We are here in the [`getPackagesToRelease`](https://github.com/changesets/changesets/blob/8be74293bb8c991d403058782d76849789dbdef9/packages/cli/src/commands/add/createChangeset.ts#L38) that is called by [`createChangeset`](https://github.com/changesets/changesets/blob/8be74293bb8c991d403058782d76849789dbdef9/packages/cli/src/commands/add/createChangeset.ts#L117), which in turn is called by [`add`](https://github.com/changesets/changesets/blob/8be74293bb8c991d403058782d76849789dbdef9/packages/cli/src/commands/add/index.ts#L18). Very similar things are already computed in `add` ([packages](https://github.com/changesets/changesets/blob/8be74293bb8c991d403058782d76849789dbdef9/packages/cli/src/commands/add/index.ts#L23), [changedPackagesName](https://github.com/changesets/changesets/blob/8be74293bb8c991d403058782d76849789dbdef9/packages/cli/src/commands/add/index.ts#L40)) - can't we trust those inputs? Why are we recomputing this stuff here? 🤔

JakeGinnivan replied:
💯

Andarist commented on packages/cli/src/commands/add/createChangeset.ts:
```diff
@@ -58,6 +60,16 @@ async function getPackagesToRelease(
+    isListablePackage(config, pkg.packageJson)
+  );
+  changedPackages = changedPackages.filter(
+    pkgName => !isListablePackage(config, pkgJsonsByName.get(pkgName)!)
```
shouldnt this check if a package is listable and not if the package is not listable?
```suggestion
    pkgName => isListablePackage(config, pkgJsonsByName.get(pkgName)!)
```

Andarist commented on packages/cli/src/commands/publish/getUntaggedPrivatePackages.ts:
```diff
@@ -0,0 +1,34 @@
+      untagged.push({
+        name: packageWithTag.pkg.packageJson.name,
+        newVersion: packageWithTag.pkg.packageJson.version,
+        published: false
```
nit: as long as we are not using this property - I would drop it entirely from here. It has a constant value for all possible items returned from this function and thus it looks a little bit misleading. When I see `published: false` I somewhat expect that `published: true` is also a valid result - but in reality that second case doesn't happen in practice.

JakeGinnivan replied:
There is logic filtering on this elsewhere, I think i'd prefer to leave it for now

Andarist commented on packages/cli/src/commands/publish/getUntaggedPrivatePackages.ts:
```diff
@@ -0,0 +1,34 @@
+) {
+  const packageWithTags = await Promise.all(
+    privatePackages.map(async privatePkg => {
+      const tagName = `${privatePkg.packageJson.name}@${privatePkg.packageJson.version}`;
```
We also support single package repositories and they use a different tag scheme. Having a single private package probably should probably be supported too

JakeGinnivan replied:
Ok, so this has highlighted a small issue, which I don't think is worth resolving atm.

Since this PR was opened the --no-git-tag flag has been added. This approach requires git tags to work, so I am going to continue to ignore that flag until we have some other state somewhere to track private package publishing (the RFCs should flush that out).

JakeGinnivan replied:
Updated to support

Andarist commented on packages/cli/src/commands/publish/publishPackages.ts:
```diff
@@ -88,18 +88,21 @@ export default async function publishPackages({
-  if (unpublishedPackagesInfo.length === 0) {
-    warn("No unpublished packages to publish");
-  }
+  const twoFactorState: TwoFactorState =
```
is this equivalent to returning `[]` when `unpublishedPackagesInfo.length === 0`?

JakeGinnivan replied:
Yep, much cleaner. Nice

Andarist commented on packages/config/src/index.ts:
```diff
@@ -462,7 +462,16 @@ export let parse = (json: WrittenConfig, packages: Packages): Config => {
     },
+
+    // TODO consider enabling this by default in the next major version
+    privatePackages: json.privatePackages
```
could we throw an error for `{ version: false, tag: true }`? this is an impossible combination to handle, right?

Andarist commented on packages/config/schema.json:
```diff
@@ -75,6 +75,20 @@
+          }
+        },
+        {
+          "type": "boolean"
```
what does `privatePackages: true` mean? should it enable all features? either way - i think it's not handled right now, see my comment here: https://github.com/changesets/changesets/pull/662/files#r983293021

Andarist commented on packages/types/src/index.ts:
```diff
@@ -61,13 +61,20 @@ export type PackageGroup = ReadonlyArray<string>;
   access: AccessType;
   baseBranch: string;
+  /** Features enabled for Private packages */
+  privatePackages: false | PrivatePackages;
```
I think you have meant to have this here:
```suggestion
  privatePackages: {
    version: boolean;
    tag: boolean;
  };
```

The `Config` type represents the already normalized config.

Andarist commented on packages/types/src/index.ts:
```diff
@@ -90,6 +97,11 @@ export type WrittenConfig = {
   access?: AccessType;
   baseBranch?: string;
+  /** Opt in to tracking non-npm / private packages */
+  privatePackages?: {
+    version?: boolean;
+    tag?: boolean;
+  };
```
OTOH, in here you have meant this:
```suggestion
  privatePackages?: false | {
    version?: boolean;
    tag?: boolean;
  };
```

Note that we should either also support `true` or change the schema (see the other comment [here](https://github.com/changesets/changesets/pull/662/files#r983293668))

Andarist replied:
The `WrittenConfig` also accepts `false`
```suggestion
  } | false;
```

JakeGinnivan commented on packages/config/src/index.ts:
```diff
@@ -462,7 +462,16 @@ export let parse = (json: WrittenConfig, packages: Packages): Config => {
+    // TODO consider enabling this by default in the next major version
+    privatePackages: json.privatePackages
+      ? {
+          version: json.privatePackages.version ?? true,
+          tag: json.privatePackages.tag ?? false,
```
🤦‍♂️

Andarist commented on packages/config/src/index.ts:
```diff
@@ -462,7 +462,29 @@ export let parse = (json: WrittenConfig, packages: Packages): Config => {
   };
+
+  if (
+    config.privatePackages &&
```
nit: in the normalized config this is always present
```suggestion
```

mino01x commented on packages/cli/src/commands/add/isListablePackage.ts:
```diff
@@ -0,0 +1,17 @@
+    return false;
+  }
+
+  if (!config.privatePackages && packageJson.private) {
```
is something wrong here? because `config.privatePackages` is always `Object`
```js
!config.privatePackages?.version && packageJson.private
```
or i misunderstand?

JakeGinnivan replied:
@mino01x is it causing an issue? If it is then we need to write a test around it

mino01x replied:
#985 yes i have create a pr. you can help me to check it.

chathu-novade commented on packages/cli/src/commands/add/isListablePackage.ts:
```diff
@@ -0,0 +1,17 @@
+import { Config } from "@changesets/types";
+import { PackageJSON } from "@changesets/types";
```
`import type { PackageJSON } from "@changesets/types";`

I think both import can be type import
